### ISSUE TITLE
Multiple performance issues

### ISSUE THREAD

**Issue by ndobromirov** [codemix/yii2-streamlog#2]
I needed a streamable logger to be used in Heroku with Yii and I considered this one, but It is exhibiting multiple issues in relation to the logging implementation itself.

## Issues found:
1. Configuration is validated on every export() call.
2. On every export(), it is creating a new file handle to the stream the logs and destroys it. That's a HUGE overhead for cases like - write every log line. 
3. Invalid / non-recommended behavior for CLI context execution. For console scripts, there are already defined constants with open streams for I/O, so the **fopen** and **fclose** are total overhead. More details [here](http://php.net/manual/en/features.commandline.io-streams.php).

## Proposed solutions.
1. Validate configurations **only once** in __construct() or **preferably** init().
2. Decide on execution context (CLI / other) on init. If not CLI - open a new handle, close it on __destruct() for.
3. For CLI execution use the STDOUT and STDERR constants, without reopening them or closing them on __destruct().

## Final thoughts.
This way the logging action will do ONLY what's important there - actual logging.

**Comment by mikehaertl** [comment]
@ndobromirov Thanks for your ideas, seems you spent quite some time on this ;). I'm a bit hesitant, though: Your proposed changes introduce quite some level of complexity to an otherwhise very simple and easy to grasp extension.

About the issues you raised:

1.) and 2.): I've followed the same logic that other Yii log targets use. I've never considered having so many calls to `export()` anyway, as usually it only gets called after a request ends or when 1000 messages are reached. On the other hand, sending every message out immediately can also be very helpful, especially in a container context. So you have set `flushInterval = 1` I assume?

3.) Hmm, isn't this an alternative fix for keeping a list of open stream handles? At least for the standard CLI streams?

My conclusion so far: I see your point and while it first looks like a very special use case it makes a lot of sense in cloud/container hosting scenarios - and that was the main motivation for this library in the first place.

But what I don't really like: The changes introduce complexity and make it hard to follow the main logic.

I wonder if you really every use any other stream than `STDIN` / `STDOUT` where the problems you raised matter. Couldn't we simply create another class like `CliTarget` or similar where we use the defined constants? And keep the `Target` unchanged?

**Comment by ndobromirov** [comment]
Yes I was considering to use **flushInterval = 1** for local development and debugging trace level logs in local docker containers.

I can agree with most of your comments. Here are 2 new PRs that make effectively the same thing, with respect to the fact that #4 has everything in a single class and #5 separates that in 2, making it more decoupled and maintainable (DRY)

**Comment by mikehaertl** [comment]
@ndobromirov I've adopted your changes in a commit to master. Can you help testing please?

**Comment by mikehaertl** [comment]
@ndobromirov There's now one new drawback: The `fp` is always opened, even if there's nothing to log in a request at all. We could fix this by moving the stuff that is now in `init()` into a getter/setter `getFp()` / `setFp()` and call `getFp()` in `export()`.

I'd say it's not really worth it - on the other hand, as we speak about performance - we could optimize it to the max ;). What do you think?

**Comment by mikehaertl** [comment]
@ndobromirov I've pushed two more changes. Maybe you can have another look. It should now really be solid, I hope. If you confirm, I'll draft a `1.2.0` release.

**Comment by ndobromirov** [comment]
Both ways are ok. I tend to go to the lazy approach with the getter and setter, as 1 I/O is always slower than many conditionals :) or plain PHP function calls.

**Comment by ndobromirov** [comment]
Seems ok. I've tested it - working locally.

**Comment by mikehaertl** [comment]
Great, 1.2.0 is out! Thanks for your help!

**Comment by ndobromirov** [comment]
Small thing that I've just found - STDOUT and STDER are in binary write mode (wb).
A tweak is needed to make it work there from
```
if ($metadata['mode']!=='w') {
```
to
```
if (!in_array($metadata['mode'], ['w', 'wb', 'a', 'ab'])) {
```

Consider opening files in append mode, instead of w, as they are re-written, so on each request logs are lost (when stream is not some pre-defined one). Or expose the mode for configuration. :)

Should this be a new issue or we can reopen this one?
This is affecting only the new feature with the stream constants.

**Comment by mikehaertl** [comment]
@ndobromirov Pushed - can you verify?

**Comment by ndobromirov** [comment]
I was using the proposed fix, before posting it - confirm that it works. Closing.
Tested the package with tag: dev-master - working.

**Comment by mikehaertl** [comment]
@ndobromirov Coming back to this issue with a question: Could you actually make immediate logging work? I've now tried with `'flushInterval' => 1` and some `sleep(5)`. Still all log output is only sent to syslog after the request ends. I wonder if the stream is buffered in some way and maybe needs to be flushed after each write operation. Couldn't really find something. Any hints?

**Comment by mikehaertl** [comment]
@ndobromirov Found it: I forgot to set `exportInterval` to 1 on the log target. For the record, here's an example config for immediate logging:

```
        'log' => [
            'flushInterval' => 1,
            'targets' => [
                [
                    'class' => 'codemix\streamlog\Target',
                    'url' => 'php://stdout',
                    'exportInterval' => 1,
                    'levels' => ['info','trace'],
                    'logVars' => [],
                    'replaceNewline' => '',
                ],
                [
                    'class' => 'codemix\streamlog\Target',
                    'url' => 'php://stderr',
                    'exportInterval' => 1,
                    'levels' => ['error', 'warning'],
                    'logVars' => [],
                    'replaceNewline' => '',
                ],
            ],
        ],
```

**Comment by mikehaertl** [comment]
@ndobromirov Final thought: I did some tests to find out about real performance implications. I therefore did several runs with this code using the configuration above.

```php
        $time = microtime(true);
        for ($x=0; $x < 100; $x++) {
            Yii::info('Dummy', __METHOD__);
        }
        $diff = round((microtime(true) - $time) * 1000, 2);
        Yii::info("Spent $diff ms", __METHOD__);
```

The results vary a lot:

| Version | Time spent |
| ----- | ----- |
| 1.1.0 | 5 - 30 ms |
| 1.2.1 | 5 - 15 ms |

I think the real benefit depends a lot on the underlying OS and how it handles file operations. I assume that modern OSes will already optimize operations as much as possible. But still it's better to be as friendly as possible to our resources, so the changes we did here still make sense.

**Comment by ndobromirov** [comment]
It does, but in cloud IOPs are deadly slow :) and you can not compare your local SSD against that. It's like apples to oranges... Either way the new version **IS** 2 times faster in the worst case (flush each log) and much faster in the no-log case, as no IOPs are going to be performed :).

The numbers on your test seem too slow (even for magnet disks). How are you testing that? Through web, console, plain script? The snippet is showing a hard dependency to the Yii framework, so this can not be the whole bench script.

To reduce the variation, have the script run more iterations (10-100 k). Also have it tested and aggregate the end results after 5-10 executions to reduce noise from other processes, that caused the variation that you've mentioned.

**Comment by mikehaertl** [comment]
I had the loop in a controller action and hit reload a couple of times. So of course, it's not very representative.

The numbers are from one of my dev servers at Hetzner. It's the smallest vServer available there: https://www.hetzner.de/gb/hosting/produkte_vserver/cx10. Not sure if my server already has SSD, could still be magnetic disks (from an older package they offered). As the machine is shared with other customers there, it explains the variations.

As said before: I have no doubt, that the changes make sense. So all good :)
